### ISSUE TITLE
Every file has executable bit

### ISSUE THREAD

**Issue by voxik** [lsegal/yard#1147]
Not sure what went wrong with 0.9.12 release, but every file in that package has executable bit set:

~~~
$ gem unpack yard
Fetching: yard-0.9.12.gem (100%)
Unpacked gem: '/home/vondruch/yard/yard-0.9.12'

$ ll yard-0.9.12
total 84
-rwxrwxrwx.  1 vondruch vondruch 28307 Dec  1 18:27 CHANGELOG.md
-rwxrwxrwx.  1 vondruch vondruch  2979 Dec  1 18:27 LEGAL
-rwxrwxrwx.  1 vondruch vondruch  1060 Dec  1 18:27 LICENSE
-rwxrwxrwx.  1 vondruch vondruch 13068 Dec  1 18:27 README.md
-rwxrwxrwx.  1 vondruch vondruch  1052 Dec  1 18:27 Rakefile
drwxrwxr-x.  2 vondruch vondruch  4096 Dec  1 18:27 benchmarks
drwxrwxr-x.  2 vondruch vondruch  4096 Dec  1 18:27 bin
drwxrwxr-x.  4 vondruch vondruch  4096 Dec  1 18:27 docs
drwxrwxr-x.  3 vondruch vondruch  4096 Dec  1 18:27 lib
drwxrwxr-x. 14 vondruch vondruch  4096 Dec  1 18:27 spec
drwxrwxr-x.  4 vondruch vondruch  4096 Dec  1 18:27 templates
-rwxrwxrwx.  1 vondruch vondruch  1808 Dec  1 18:27 yard.gemspec


$ gem unpack yard -v 0.9.11
Fetching: yard-0.9.11.gem (100%)
Unpacked gem: '/home/vondruch/yard/yard-0.9.11'

$ ll yard-0.9.11
total 84
-rw-r--r--.  1 vondruch vondruch 28094 Dec  1 18:28 CHANGELOG.md
-rw-r--r--.  1 vondruch vondruch  2979 Dec  1 18:28 LEGAL
-rw-r--r--.  1 vondruch vondruch  1060 Dec  1 18:28 LICENSE
-rw-r--r--.  1 vondruch vondruch 13180 Dec  1 18:28 README.md
-rw-r--r--.  1 vondruch vondruch  1052 Dec  1 18:28 Rakefile
drwxrwxr-x.  2 vondruch vondruch  4096 Dec  1 18:28 benchmarks
drwxrwxr-x.  2 vondruch vondruch  4096 Dec  1 18:28 bin
drwxrwxr-x.  4 vondruch vondruch  4096 Dec  1 18:28 docs
drwxrwxr-x.  3 vondruch vondruch  4096 Dec  1 18:28 lib
drwxrwxr-x. 14 vondruch vondruch  4096 Dec  1 18:28 spec
drwxrwxr-x.  4 vondruch vondruch  4096 Dec  1 18:28 templates
-rw-r--r--.  1 vondruch vondruch  1808 Dec  1 18:28 yard.gemspec
~~~

The repository looks to be OK.

**Comment by lsegal** [comment]
Thanks for the report. This looks like an issue with releasing from a Windows machine.

**Comment by idima89** [comment]
Still actual with yard-0.9.14:

```
16:13:09 [dima@thinkpad ~/.bundle/ruby/2.3.0/gems/yard-0.9.14]$ find -type f -perm 777 | wc -l
578
16:14:36 [dima@thinkpad ~/.bundle/ruby/2.3.0/gems/yard-0.9.14]$ ls -l
total 72
drwxr-xr-x  2 dima dima  4096 Jun 15 22:38 benchmarks
drwxr-xr-x  2 dima dima    43 Jun 15 22:38 bin
-rwxrwxrwx  1 dima dima 28825 Jun 15 22:38 CHANGELOG.md
drwxr-xr-x  4 dima dima   212 Jun 15 22:38 docs
-rwxrwxrwx  1 dima dima  2979 Jun 15 22:38 LEGAL
drwxr-xr-x  3 dima dima    59 Jun 15 22:38 lib
-rwxrwxrwx  1 dima dima  1060 Jun 15 22:38 LICENSE
-rwxrwxrwx  1 dima dima  1052 Jun 15 22:38 Rakefile
-rwxrwxrwx  1 dima dima 13068 Jun 15 22:38 README.md
drwxr-xr-x 14 dima dima  4096 Jun 15 22:38 spec
drwxr-xr-x  4 dima dima    34 Jun 15 22:38 templates
-rwxrwxrwx  1 dima dima  1808 Jun 15 22:38 yard.gemspec
```

**Comment by lsegal** [comment]
Marking this as closed since there's no way to fix this in old releases.

**Comment by idima89** [comment]
Hi, @lsegal. Which release of yard is new without this issue? I see yard-0.9.15 on rubygems.org, which is still affected:
```
23:42:26 [dima@thinkpad ~/.bundle/ruby/2.3.0/gems/yard-0.9.15]$ find -type f -perm 777 | wc -l
579
23:44:18 [dima@thinkpad ~/.bundle/ruby/2.3.0/gems/yard-0.9.15]$ ls -l
total 72
drwxr-xr-x  2 dima dima  4096 Aug  2 17:55 benchmarks
drwxr-xr-x  2 dima dima    43 Aug  2 17:55 bin
-rwxrwxrwx  1 dima dima 29101 Aug  2 17:55 CHANGELOG.md
drwxr-xr-x  4 dima dima   212 Aug  2 17:55 docs
-rwxrwxrwx  1 dima dima  2979 Aug  2 17:55 LEGAL
drwxr-xr-x  3 dima dima    59 Aug  2 17:55 lib
-rwxrwxrwx  1 dima dima  1060 Aug  2 17:55 LICENSE
-rwxrwxrwx  1 dima dima  1052 Aug  2 17:55 Rakefile
-rwxrwxrwx  1 dima dima 13068 Aug  2 17:55 README.md
drwxr-xr-x 14 dima dima  4096 Aug  2 17:55 spec
drwxr-xr-x  4 dima dima    34 Aug  2 17:55 templates
-rwxrwxrwx  1 dima dima  1808 Aug  2 17:55 yard.gemspec
```

**Comment by lsegal** [comment]
There is no current release that fixes this. This issue would be fixed in the next release if at all. Can you provide more information about how this affects the behavior of the library?

**Comment by pegasd** [comment]
I'm somewhat disappointed at `if at all` statement. Is there no way to prevent this in new releases?
This is not a "behavior" issue, as you say, but a security concern.

**Comment by lsegal** [comment]
Unfortunately it's difficult to prevent this when releasing from a Windows machine via [WSL](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux) because Windows itself doesn't really have a concept of executable bits or chmod permissions in the same way that Linux does. There's a pretty high burden on guaranteeing consistency of these releases when using standard Ruby tools. Releases seem to work OK on normal Windows, probably due to some assumptions RubyGems makes when packaging, but fail via WSL. It might be something RubyGems wants to handle, or not, but my current release process occasionally requires releasing via WSL for consistency of tooling and other package tasks (we package via Samus under a WSL environment on Windows to ensure that all tooling is available). Releases used to be done on macOS but I moved away from doing those there due to other constraints. They still can be if needed, but this is the "if at all" part: I cannot guarantee that all releases will be done via a macOS system.

I may attempt to revisit some of the tooling / package steps on WSL to try to work around this, but if I'm understanding that this is a security issue regarding world rwx permissions, I would defer to RubyGems for handling this. You're right that YARD shouldn't be doing this, but realistically, _no_ gem shouldn't be able to install files with world permissions on a third-party system. In my opinion, that's something RubyGems should be taking on as a security issue in `gem install` or even `gem build`. It seems like this was sort-of brought up in a few discussions in rubygems/rubygems#1041 but it doesn't look like it's gone anywhere. I would recommend pushing on that issue to ensure that other libraries can't accidentally fall into the same trap. If we got that into RubyGems, this wouldn't be a YARD issue anymore, and IMO, is better ergonomics for the community as a whole rather than treating this as one-off packaging issues.
